init file download via POST

blizzz opened the pull request:
Currently file downloads are triggered by a regular GET request through changing the window location. This can leak file names, but also limits the number of downloadable files for URL length restrictions. 

~~By switching to a POST request (thus through XHR) we can avoid both aspects. The downside with the chosen method is, that it does not stream the data: the UI is not blocked, but the browser's download progress does not kick in and the save dialog appears only after the file is downloaded. ~~

* [X] Hence, there is not visual feedback. 

Switch to register the download action via POST call that provides all the information. A token is returned, which is then passed along by the GET request through location change. Save dialogue is shown immediately and the browser progress indicator works again, too.

As far as I researched there is no way around with XHR (no stream support by design). Streams API offer support, alas – as far as i know - still would suffer the same problems around the built-in progress bar and save dialog (and cannot be backported for lack of IE support).

~~Another approach would be to use the progress indicator as with file uploads with downloads, too. But maybe I also oversaw an option?~~

Now:

* [ ] download on sharing pages is broken
* [ ] tests
* and the open review remarks

blizzz committed:
Download files via POST request

- also turns old download php script to Controller
- increases number of possible files for lifting length restrictions in GET data

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

skjnldsv commented on apps/files/src/services/Download.js:
```diff
@@ -0,0 +1,52 @@
+
+export default class Download {
+
+	export
+	const
+	async get(files, dir, downloadStartSecret) {
```
No! This is not compliant to our js coding style standards :grin:

blizzz replied:
Okay. What exactly?

skjnldsv replied:
`export const get = async function(files, dir, downloadStartSecret) {`

blizzz replied:
changed that, too, is it ok now?

skjnldsv commented on apps/files/src/services/Download.js:
```diff
@@ -0,0 +1,52 @@
+	export
+	const
+	async get(files, dir, downloadStartSecret) {
+		await axios.post(
+			generateUrl('apps/files/ajax/download.php'),
+			{
+				files,
+				dir,
+				downloadStartSecret,
+			},
+			{
+				responseType: 'blob',
+			}
+		).then(res => {
```
Don't use async-await and then

blizzz replied:
Okay.

skjnldsv commented on apps/files/src/services/Download.js:
```diff
@@ -0,0 +1,52 @@
+import { generateUrl } from '@nextcloud/router'
+import axios from '@nextcloud/axios'
+
+export default class Download {
```
Doesn't need to be a class, you can export de function directly

blizzz replied:
Yes. I thought it's more flexible if it should ever be extended, but maybe over engineered.

skjnldsv replied:
No, it's not, if it should be extended, we can just add new exports. They are modules :)

You can put as many `export const xxx = function() {` :wink:

blizzz replied:
Turned it into  a function, this is how you like it now?

skjnldsv commented on apps/files/src/download.js:
```diff
@@ -0,0 +1,29 @@
+if (!window.OCA.Files) {
+	window.OCA.Files = {}
+}
+Object.assign(window.OCA.Files, { Download: new Download() })
```
Import the function directly and pass it as a reference instead of using `new`

blizzz replied:
Okay. (I took example of the several other places where we do it like that)

skjnldsv replied:
Can you point out where we do that? :)

blizzz replied:
```
$ ack 'Object\.assign.*new'  apps/*/src/
apps/files_sharing/src/files_sharing_tab.js
35:Object.assign(window.OCA.Sharing, { ShareSearch: new ShareSearch() })
36:Object.assign(window.OCA.Sharing, { ExternalLinkActions: new ExternalLinkActions() })
37:Object.assign(window.OCA.Sharing, { ShareTabSections: new TabSections() })

apps/files/src/download.js
29:Object.assign(window.OCA.Files, { Download: new Download() })

apps/files/src/sidebar.js
36:Object.assign(window.OCA.Files, { Sidebar: new Sidebar() })

apps/files/src/files-app-settings.js
35:Object.assign(window.OCA.Files, { Settings: new Settings() })
```

blizzz commented:
An alternative that would keep the browser's download progress would be to announce the download with a post-request, the returns a token, and then trigger the download again by GET but only providing the token as parameter.

skjnldsv commented:
> the browser's download progress does not kick in and the save dialog appears only after the file is downloaded.

For me that is a huge issue.
If we do this, like big hosting companyes are doing (mega was one of the first), we need to show a progress bar. What if the file is 20GB, will I have to wait for 1h to see the file save dialog?

PVince81 commented:
> An alternative that would keep the browser's download progress would be to announce the download with a post-request, the returns a token, and then trigger the download again by GET but only providing the token as parameter.

Somehow reminds me of pre-signed download URLs, which are a slightly different thing though.

blizzz committed:
switch to registration via POST

- download relevant data is sent per POST, a token is received
- download is started by GET and the received token
- solves the disadvantages from XHR-only download
- job to cleanup download tokens

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
> An alternative that would keep the browser's download progress would be to announce the download with a post-request, the returns a token, and then trigger the download again by GET but only providing the token as parameter.

This is now implemented.  

On smoke testing it of course turned out that sharing needs extra care. There is duplication to some degree, so that's still to tackle.

blizzz committed:
(WIP) consolidate download utilities

- move registration and cleanup code to new DownloadManager
- adjust files and files_sharing controllers accordingly
- TODO: js code of files_sharing
- TODO: tests

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
register alias for DownloadManager

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
turn download class into a function

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

skjnldsv commented on apps/files/src/download.js:
```diff
@@ -0,0 +1,28 @@
+ *
+ */
+
+import download from './services/Download'
```
I would argue that the service file name is a bit too simple and confusing.
But it's no blocker: 

Something like ?
```suggestion
import download from './services/DownloadFileService'
```

blizzz replied:
Okay. Just, service is now duplicated (in the path), and what else do you download other than files? It's also actually executing a download. You could argue whether the business logic (register first, then execute), however I think that's an implementation detail here.

skjnldsv commented on apps/files/src/services/Download.js:
```diff
@@ -0,0 +1,43 @@
+import { generateUrl } from '@nextcloud/router'
+import axios from '@nextcloud/axios'
+
+export default function(files, dir, downloadStartSecret) {
+	axios.post(
+		generateUrl('apps/files/registerDownload'),
+		{
+			files,
+			dir,
+			downloadStartSecret,
+		}
+	).then(res => {
+		if (res.status === 200 && res.data.token) {
+			const dlUrl = generateUrl(
+				'apps/files/ajax/download.php?token={token}',
+				{ token: res.data.token }
+			)
+			OC.redirect(dlUrl)
+		}
+	})
+}
```
```suggestion
export default async function(files, dir, downloadStartSecret) {
	await res = axios.post(generateUrl('apps/files/registerDownload'), {
		files,
		dir,
		downloadStartSecret,
	})

	if (res.status === 200 && res.data.token) {
		const dlUrl = generateUrl('apps/files/ajax/download.php?token={token}',{
			token: res.data.token
		}
		OC.redirect(dlUrl)
	}
}
```

skjnldsv replied:
It's cleaner to use async/await than then/catch

```suggestion
export default async function(files, dir, downloadStartSecret) {
	const res = await axios.post(generateUrl('apps/files/registerDownload'), {
		files,
		dir,
		downloadStartSecret,
	})

	if (res.status === 200 && res.data.token) {
		const dlUrl = generateUrl('apps/files/ajax/download.php?token={token}',{
			token: res.data.token
		}
		OC.redirect(dlUrl)
	}
}
```

blizzz commented on apps/files_sharing/lib/Controller/ShareController.php:
```diff
@@ -538,33 +550,58 @@ public function showShare($path = ''): TemplateResponse {
 	 * @throws NotFoundException
 	 */
-	public function downloadShare($token, $files = null, $path = '', $downloadStartSecret = '') {
+	public function downloadShare(string $downloadToken) {
```
It's also debatable to keep this stable, definitely not to backport it. It might be the case that this is used by other tools (or scripts) in the wild to download files from shares.

blizzz committed:
prefer async/await over then

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

Co-authored-by: John Molakvoæ <skjnldsv@users.noreply.github.com>

blizzz commented:
unfinished and not and the table, closing.
